### PR TITLE
render waypoint distance in the world

### PR DESCRIPTION
**Pull request by HalcyonSuoh**

### CODE REVIEW

File: common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    @Config
+    public CustomColor textColor = CommonColors.WHITE;
+
+    @Config
+    public int backgroundOpacity = 80;
+
+    @Config
+    public FontRenderer.TextShadow textShadow = FontRenderer.TextShadow.NONE;
+
+    private Vector3d ndc = null;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java]: Can you rename this to be more specific, so anyone reading can understand what this stores?

File: common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+    @SubscribeEvent
+    public void onRenderLevelPost(RenderLevelEvent.Post event) {
+        if (CompassModel.getCompassLocation().isEmpty()) return;
+
+        Location location = CompassModel.getCompassLocation().get();
+        Matrix4f projection = event.getProjectionMatrix().copy();
+        Camera camera = event.getCamera();
+        Vec3 cameraPos = camera.getPosition();
+        Vector3f positiveX = new Vector3f(1.0f, 0.0f, 0.0f);
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java]: Vector3f.XP

File: common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        Vector3f positiveY = new Vector3f(0.0f, 1.0f, 0.0f);
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java]: Vector3f.YP

**Review by kristofbolyai — COMMENTED** [review]
I will finish reviewing the rest of the changes later.

File: common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                        textColor,
+                        HorizontalAlignment.Left,
+                        textShadow);
+    }
+
+    // ndc range check, make sure the point is on screen
+    // z check is flipped to countereact matrix transformation in onRenderGuiPost
+    // possible behavior change?
+    private boolean isOnScreen() {
+        return !(ndc.x > 1 || ndc.x < -1 || ndc.y > 1 || ndc.y < -1 || !(ndc.z > 1));
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java]: Bit weird, you could flip the whole thing to remove `!`?

File: common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                displayPositionX - (backgroundHeight / 2) - 2,
+                displayPositionY - 2,
+                0,
+                backgroundWidth + 3,
+                backgroundHeight + 2);
+        FontRenderer.getInstance()
+                .renderAlignedTextInBox(
+                        event.getPoseStack(),
+                        distanceText,
+                        displayPositionX - (backgroundHeight / 2),
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java]: Use the `renderAlignedTextInBox` which can align vertically and horizontally?

File: common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java (hunk 4 above)
Comment (HalcyonSuoh) [reply common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java]: the z check is a somewhat specific thing, I fear flip the whole thing might be unintutive, and possibly easier to get overlooked when glancing over

File: common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java (hunk 1 above)
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java]: @HalcyonSuoh I dont think you renamed the variable?

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        float displayPositionX = (float) ((ndc.x + 1.0f) / 2.0f) * window.getGuiScaledWidth();
+        float displayPositionY = (float) ((1.0f - ndc.y) / 2.0f) * window.getGuiScaledHeight();
+
+        RenderUtils.drawRect(
+                event.getPoseStack(),
+                CommonColors.BLACK.withAlpha(backgroundOpacity),
+                displayPositionX - (backgroundWidth / 2) - 2,
+                displayPositionY - (backgroundHeight / 2) - 2,
+                0,
+                backgroundWidth + 3,
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java:103]: Shouldn't you do + 2 to be in-line with every other alignment?

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                        0,
+                        textColor,
+                        HorizontalAlignment.Center,
+                        VerticalAlignment.Middle,
+                        textShadow);
+    }
+
+    // ndc range check, make sure the point is on screen
+    // z check is flipped to countereact matrix transformation in onRenderGuiPost
+    // possible behavior change?
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java]: Either remove this comment or make it more understandable

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    private boolean isOnScreen() {
+        return !(ndc.x > 1 || ndc.x < -1 || ndc.y > 1 || ndc.y < -1 || !(ndc.z > 1));
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java]:
```suggestion
        return ndc.x <= 1 && ndc.x >= -1 && ndc.y <= 1 && ndc.y ?= -1 && ndc.z > 1;
```

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+import net.minecraft.world.phys.Vec3;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class WorldWaypointDistanceFeature extends UserFeature {
+
+    @Config
+    public CustomColor textColor = CommonColors.WHITE;
+
+    @Config
+    public int backgroundOpacity = 80;
```
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java]: I would rather you make this a float (0-1 value). I think every other opacity config is a float.

**Review by kristofbolyai — COMMENTED** [review]
Looks better! Only a few comments + rename ndc, and we are good to go :)

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java (hunk 6 above)
Comment (HalcyonSuoh) [reply common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java:103]: image
+2 doesn't seems to be enough space, make the m looks like it's suffocating, kerning might be the cause of this, but I don't think editing that is something I can or should do

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java:103]: We probably lose 1 pixel in some floating point to int conversion or smth. I guess keep it.

File: common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java (hunk 1 above)
Comment (HalcyonSuoh) [reply common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java]: I failed to find a more descriptive name myself, looking around codebase only to find other TextShadow is also called textShadow, `AuraTimerOverlayFeature` for instance.
as for ndc, which stands for Normalized Device Coordinates, which I also failed to find other appropriate name for, ditching the abbreviation make the variable name way longer than it should be, and I don't think I can think of something else by myself

Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java]: textShadow is fine, I didn't want you change that. But for ndc, I would rather have a long name, than a 3 char one.

Comment (HalcyonSuoh) [reply common/src/main/java/com/wynntils/features/user/BeaconDistanceFeature.java]: I still doubt anyone seeing Normalized Device Coordinates know what it is or what it does, but I guess that make google experience smoother than the abbreviation

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+    @SubscribeEvent
+    public void onRenderGuiPost(RenderEvent.Post event) {
+        if (CompassModel.getCompassLocation().isEmpty() || normalizedDeviceCoordinates == null || isOnScreen()) return;
+
+        Window window = event.getWindow();
+
+        float backgroundWidth = FontRenderer.getInstance().getFont().width(distanceText);
+        float backgroundHeight = FontRenderer.getInstance().getFont().lineHeight;
+        WynntilsMod.info(String.valueOf(backgroundWidth));
```
Comment (magicus) [inline common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java]: Debug code left over?

File: common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+    private boolean isOnScreen() {
+        return normalizedDeviceCoordinates.x <= 1
+                && normalizedDeviceCoordinates.x >= -1
+                && normalizedDeviceCoordinates.y <= 1
+                && normalizedDeviceCoordinates.y >= -1
+                && normalizedDeviceCoordinates.z > 1;
+    }
+
+    private void worldToNdc(Vector3f delta, Matrix4f projection) {
```
Comment (magicus) [inline common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java]: This side-effect coding makes the function very opaque. Better if it returns a Vector3d, and you make the assignment at the call site.

**Review by magicus — CHANGES_REQUESTED** [review]

**Review by magicus — APPROVED** [review]

**Review by kristofbolyai — APPROVED** [review]

**Pull request merged**